File: mobile/android/crypto/vault/AttachmentCrypto.kt
package com.speakeasy.crypto.vault

import android.util.Base64
import java.security.MessageDigest
import javax.crypto.Cipher
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec
import java.security.SecureRandom

// Encryption/decryption for attachments using AES-GCM

object AttachmentCrypto {
    
    // Encrypt data with key
    fun encrypt(plaintext: ByteArray, key: SecretKey): Pair<ByteArray, ByteArray> {
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, key)
        val nonce = cipher.iv
        val ciphertext = cipher.doFinal(plaintext)
        return Pair(ciphertext, nonce)
    }
    
    // Decrypt data with key and nonce
    fun decrypt(ciphertext: ByteArray, nonce: ByteArray, key: SecretKey): ByteArray {
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, key, GCMParameterSpec(128, nonce))
        return cipher.doFinal(ciphertext)
    }
    
    // Compute SHA256 hash (Base64)
    fun sha256Base64(data: ByteArray): String {
        val digest = MessageDigest.getInstance("SHA-256")
        return Base64.encodeToString(digest.digest(data), Base64.NO_WRAP)
    }
}

File: mobile/ios/SpeakeasyCrypto/Vault/AttachmentCrypto.swift
import Foundation
import CryptoKit

// Streaming encryption/decryption for attachments using AES-GCM

class AttachmentCrypto {
    
    // Encrypt data with a provided key
    static func encrypt(plaintext: Data, key: SymmetricKey) throws -> (ciphertext: Data, nonce: Data) {
        let nonce = AES.GCM.Nonce()
        let box = try AES.GCM.seal(plaintext, using: key, nonce: nonce)
        
        return (ciphertext: box.ciphertext + box.tag, nonce: Data(nonce))
    }
    
    // Decrypt data with a provided key and nonce
    static func decrypt(ciphertext: Data, nonce: Data, key: SymmetricKey) throws -> Data {
        let gcmNonce = try AES.GCM.Nonce(data: nonce)
        
        // Separate ciphertext and tag (last 16 bytes)
        let tagLength = 16
        let encryptedData = ciphertext.prefix(ciphertext.count - tagLength)
        let tag = ciphertext.suffix(tagLength)
        
        let box = try AES.GCM.SealedBox(nonce: gcmNonce, ciphertext: encryptedData, tag: tag)
        return try AES.GCM.open(box, using: key)
    }
    
    // Compute SHA256 hash of data (for ciphertext verification)
    static func sha256(data: Data) -> String {
        let hash = SHA256.hash(data: data)
        return Data(hash).base64EncodedString()
    }
}
